appvia/kore
# Support template parameters in service plans to allow to reference cluster parameters

## What

This PR allows us to dynamically refer values from clusters when creating services.

### New ProviderData field on Cluster.Status

The `ProviderData` field exposes provider specific data, e.g. AWS account ID and VPC details on the cluster objects.

Example:
```
providerData:
      awsAccountID: "1234567890"
      vpc:
        availabilityZoneIDs:
        - euw2-az1
        - euw2-az2
        - euw2-az3
        availabilityZoneNames:
        - eu-west-2c
        - eu-west-2a
        - eu-west-2b
        ipv4EgressAddresses:
        - 1.2.3.4
        privateSubnetIDs:
        - subnet-12345
        publicSubnetIDs:
        - subnet-12345
        securityGroupIDs:
        - sg-12345
        vpcID: vpc-12345
```

### Template expression support in service plans

We support simple variable references using `text/template` in schemas and in the plan configuration values.

Only service plans can contain template references and services have to contain the final values. Therefore when you want to use a service plan, you have to call the new plan details API endpoint (see below) with the right team/cluster context and use those values in the CLI/UI. This PR makes these changes.

Example from the S3 kind's schema:
```
"target_account_id":{
  "default":"{{ .cluster.status.providerData.awsAccountID }}",
  "description":"AWS Account ID to provision into",
  "title":"AWS Target Account ID",
  "type":"string",
  "const":"{{ .cluster.status.providerData.awsAccountID }}"
}
```

### Ability to override plan configuration values in open service broker providers.

A new `platConfigurationOverrides` config parameter allows to override any plan values. For the awsservicebroker we set a static configuration:

```
"*": {
  {
    Name:     "target_account_id",
    Value:    "{{ .cluster.status.providerData.awsAccountID }}",
    Const:    true,
    Required: utils.BoolPtr(true),
  },
  {
    Name:     "target_role_name",
    Value:    config.AWSIAMRoleName,
    Const:    true,
    Required: utils.BoolPtr(true),
  },
  {
    Name:     "VpcId",
    Value:    "{{ .cluster.status.providerData.vpc.vpcID }}",
    Const:    true,
    Required: utils.BoolPtr(true),
  },
},
```

This configuration will amend the JSON schemas and makes these parameters required and a constant.

### New API endpoint for getting service plan details with optional context

API endpoint: **/api/v1alpha1/serviceplans/s3-dev/details**

It replaces the old team plan details endpoint and the separate schema/credentialschema endpoints.

For the given plan it returns with the schemas (falling back to the values on the service kind if necessary) and compiles the editable parameters.

If you give it a team and cluster context: `/api/v1alpha1/serviceplans/s3-dev/details?team=a-team&cluster=dev-cluster`, then it will replace the template parameters like `{{ .cluster.status.providerData.x }}` in the schemas and in the configuration values.

### Automatically create and use an IAM role in the aws-servicebroker

There is a new parameter `awsIAMRoleName` on the aws-servicebroker config (defaults to `aws-service-broker`). The provider will automatically create this role in the same account where it's installed.

If you want to use the aws-servicebroker with multiple accounts, you have to manually create the same role in all other accounts and allow the aws-servicebroker to assume this role in all accounts.

Note : we will be able to automate this fully when we have the full AWS account management logic.

### Additional changes

 * Remove the cloud selector from the service form as we only allow cloud services in the same cloud as the cluster (at least for now)
 * change `parameterEditable` map to `editableParameters` array
 * Bugfix: API client doesn't handle query parameters correctly with custom endpoint + properly encode query params
 * Bugfix: new service plan form to initialise plan data correctly (planValues initialised empty if data is set)
 * the awsservicebroker config now expects `awsAccessKeyID` and `awsSecretAccessKey` instead of the snake_case versions. It will work with the old values, but will complain when you want to update the object.

### Testing

 * Make sure you have the AWS servicebroker installed (see doc/setting-up-aws-servicebroker.md)
 * Test cloud service, app service and cluster creation.
 * Create/manage cluster and service plans
